ts/src/base/Exchange.ts

```diff
@@ -4516,6 +4516,147 @@ export default class Exchange {
+                    // do it backwards, starting from the last
+                    // UNTIL filtering is required in order to work
+                    if (lastTimestamp !== undefined) {
+                        params['until'] = lastTimestamp - 1;
+                    }
+                    const response = await this[method] (symbol, undefined, maxEntriesPerRequest, params);
+                    const responseLength = response.length;
+                    errors = 0;
+                    result = this.arrayConcat (result, response);
+                    if (responseLength < maxEntriesPerRequest) {
+                        break;
+                    }
+                    const firstElement = this.safeValue (response, 0);
+                    lastTimestamp = this.safeInteger2 (firstElement, 'timestamp', 0);
+                    if (lastTimestamp <= since) {
+                        break;
+                    }
+                } else {
+                    // do it forwards, starting from the since
+                    lastTimestamp = since;
+                    const response = await this[method] (symbol, lastTimestamp, maxEntriesPerRequest, params);
+                    const responseLength = response.length;
+                    if (responseLength < maxEntriesPerRequest) {
+                        break;
+                    }
+                    errors = 0;
+                    result = this.arrayConcat (result, response);
+                    lastTimestamp = this.safeInteger (responseLength - 1, 'timestamp');
+                }
+            } catch (e) {
+                if (errors + 1 > maxRetries) {
+                    throw e;
+                }
+                errors = errors + 1;
+            }
+        }
+        return this.removeRepeatedElementsFromArray (result);
+    }
+
+    async fetchPaginatedCallDeterministic (method: string, symbol: string = undefined, since = undefined, limit = undefined, timeframe = undefined, params = {}, maxEntriesPerRequest = 1000): Promise<any> {
+        let maxCalls = undefined;
+        [ maxCalls, params ] = this.handleOptionAndParams (params, method, 'paginationCalls', 20);
+        const now = this.milliseconds ();
+        const tasks = [];
+        const time = this.parseTimeframe (timeframe) * 1000;
+        const step = time * maxEntriesPerRequest;
+        let currentSince = now - (maxCalls * step) - 1;
+        if (since !== undefined) {
+            currentSince = Math.max (currentSince, since);
+        }
+        for (let i = 0; i < maxCalls; i++) {
+            if (timeframe && method !== 'fetchFundingRateHistory') {
+                tasks.push (this[method] (symbol, timeframe, currentSince, maxEntriesPerRequest, params));
+            } else {
+                tasks.push (this[method] (symbol, currentSince, maxEntriesPerRequest, params));
+            }
+            currentSince = currentSince + step - 1;
+        }
+        const results = await Promise.all (tasks);
```

Comment:
Should we add a log for the number of calls being made?
 if (this.verbose) {
            const numTasks = tasks.length;
            console.log ('Pagination call: making ', numTasks, ' calls' );
  }